instrumentation/java-http-client/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/javahttpclient/HttpClientInstrumentation.java

```diff
@@ -65,87 +67,114 @@ public void transform(TypeTransformer transformer) {
         HttpClientInstrumentation.class.getName() + "$SendAsyncAdvice");
   }
 
-  @SuppressWarnings("unused")
-  public static class SendAdvice {
+  public static class AdviceScope {
+    private final Context context;
+    private final Context parentContext;
+    private final Scope scope;
+    private final CallDepth callDepth;
+
+    private AdviceScope(Context parentContext, Context context, Scope scope, CallDepth callDepth) {
+      this.parentContext = parentContext;
+      this.context = context;
+      this.scope = scope;
+      this.callDepth = callDepth;
+    }
 
-    @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void methodEnter(
-        @Advice.Argument(value = 0) HttpRequest httpRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
+    @Nullable
+    public static AdviceScope start(HttpRequest request) {
+      return start(request, null);
+    }
+
+    private static AdviceScope start(HttpRequest request, CallDepth callDepth) {
       Context parentContext = currentContext();
-      if (!instrumenter().shouldStart(parentContext, httpRequest)) {
-        return;
+      if (!instrumenter().shouldStart(parentContext, request)) {
+        return null;
+      }
+
+      Context context = instrumenter().start(parentContext, request);
+      return new AdviceScope(parentContext, context, context.makeCurrent(), callDepth);
+    }
+
+    public void end(HttpRequest request, HttpResponse<?> response, Throwable throwable) {
+      scope.close();
+      instrumenter().end(context, request, response, throwable);
+    }
+
+    public static AdviceScope startWithCallDepth(HttpRequest httpRequest) {
+      CallDepth callDepth = CallDepth.forClass(HttpClient.class);
```

Comment:
You might have to double check, but can't you make `callDepth` a `static final` instance field on `AdviceScope`?

Reply:
I don't think so as the thread-local storage get is executed directly when calling `forClass()` which means it's probably safer to keep it that way.